distsqlrun: drain meta on ConsumerClosed of materializer

yuzefovich opened the pull request:
Previously, when ConsumerClosed was called on materializer, any
Inboxes that were in the chain of columnar operators under that
materializer would hang because they never received a closing
signal. Now, we will drain metadata sources in materializer
which will trigger the closure of Inboxes. This seems like a
hack to me, but it seems to work.

Addresses: #38919.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39077)
<!-- Reviewable:end -->

jordanlewis reviewed (approved):
I agree that it's a little hacky, but let's merge this. I think we should work on making flow shutdown happen in a more predictable way. It seems unsafe to me that it's so easy to let things get stuck - just one missing metadata call could cause a deadlock right now, I think.

@asubiotto is there any way we could come up with a "fail safe" way of getting flows shut down in these cases? Or is that not really possible?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39077#-:-LkZes7gEOK5IZmtqBzM:b-sd9nle)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39077#-LkZehCNEdw6ESUYtoSd:-LkZehCNEdw6ESUYtoSe:bemalo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/144f133896429d6009c173dcfc32fc2002a46d31/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L87)):*
> ```
> 
> statement ok
> SET optimizer = on
> ```

Instead of this, we should change the logic test config to run with the optimizer always on. There's no need to run logic tests with optimizer=off right now.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39077)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @solongordon)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39077#-LkZehCNEdw6ESUYtoSd:-LkZgB_FDJuEhTe21Bjx:bvkb8m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/144f133896429d6009c173dcfc32fc2002a46d31/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L87)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Instead of this, we should change the logic test config to run with the optimizer always on. There's no need to run logic tests with optimizer=off right now.
</blockquote></details>

I was wondering about the same thing. I've updated only `5node-dist` config that is used in this file, but we probably want to turn optimizer on in all of the configs.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
distsqlrun: drain meta on ConsumerClosed of materializer

Previously, when ConsumerClosed was called on materializer, any
Inboxes that were in the chain of columnar operators under that
materializer would hang because they never received a closing
signal. Now, we will drain metadata sources in materializer
which will trigger the closure of Inboxes. This seems like a
hack to me, but it seems to work.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39077)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @solongordon)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39077#-LkZehCNEdw6ESUYtoSd:-LkZoaJ7Am6b6G0aZj7p:bsebgev) ([raw file](https://github.com/cockroachdb/cockroach/blob/144f133896429d6009c173dcfc32fc2002a46d31/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L87)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I was wondering about the same thing. I've updated only `5node-dist` config that is used in this file, but we probably want to turn optimizer on in all of the configs.
</blockquote></details>

I went back to setting optimizer on manually because there were changes in planner tests. I think the change to the config should be a separate PR.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39077#-:-LkZpNBy-jdWt11U_MSj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @solongordon, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39077#-LkZehCNEdw6ESUYtoSd:-LkZpM6a274TjT_ihEHy:b-66lo4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/144f133896429d6009c173dcfc32fc2002a46d31/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L87)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I went back to setting optimizer on manually because there were changes in planner tests. I think the change to the config should be a separate PR.
</blockquote></details>

Okay, sorry.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

asubiotto commented:
#39080 seems to be related as well. I agree that we need to come up with a principled approach to shut down flows ahead of time. This is essentially what `ConsumerClosed` was (shut down without a metadata drain) and `ConsumerDone` (shut down with a metadata drain) but had to be implemented by the whole flow. I think we only have a couple of edge cases to consider (mostly components that spawn goroutines, I can think of the `UnorderedSynchronizer`, `HashRouter`, and `Inbox`). For now this is probably fine.
